src/main/resources/org/jenkinsci/plugins/workflow/cps/steps/NonCPSGlobalVariable/NonCPSWrapperImpl.groovy

```diff
@@ -0,0 +1,34 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2018 CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.jenkinsci.plugins.workflow.cps.steps.NonCPSGlobalVariable
+
+import com.cloudbees.groovy.cps.NonCPS
+
+class NonCPSWrapperImpl implements Serializable {
+
+    @NonCPS
+    <V> V exec(Closure<V> body) {
+        body()
```

Comment:
Does this actually work? I cannot imagine how it could. `@NonCPS` is applied _lexically_ to compilation of this method source. It has no effect on the CPS transform of the body closure. I think you only _think_ it works because your `printTimestamp` method returns `true` from `StepExecution.start` and thus is synchronous (would be more simply written using `SynchronousStepExecution`, and thus the CPS transformation status of the body is largely irrelevant. If you add, say, a `sleep 1` statement inside the closure I expect you will see it blow up, or rather (pending #211) quietly behave strangely.

Reply:
Will check it. in https://github.com/jenkinsci/pipeline-utility-steps-plugin/pull/46 a complex wrapper with multiple steps does not fail after such patch, but I agree it may be also not a case in real.

Reply:
Yeah, I have really mixed feelings on this. 

As a whole, the NonCPS mechanism is NOT good about advertising when it has been abused and failing appropriately -- I think @jglick might have some in-progress work that resolves that though?